lib/UnexpectedExpressMocker.js

```diff
@@ -0,0 +1,366 @@
+
+    const rawResponseChunks = [];
+    res.assignSocket(new stream.Writable());
+    res.connection._write = (chunk, encoding, cb) => {
+      rawResponseChunks.push(chunk);
+      cb();
+    };
+
+    let isDestroyed = false;
+    res.connection.destroy = () => {
+      isDestroyed = true;
+    };
+
+    let isAsync = false;
+    setImmediate(() => {
+      isAsync = true;
+    });
+
+    const context = {};
+    const nextCalls = [];
+    let done = false;
+    let errorPassedToNext;
+
+    return makePromise((resolve, reject) => {
+      ['write', 'end', 'destroy'].forEach(methodName => {
+        const orig = res[methodName];
+        res[methodName] = function(chunk, encoding) {
+          const returnValue = orig.apply(this, arguments);
+          isDestroyed = isDestroyed || methodName === 'destroy';
+          if (methodName === 'end' || methodName === 'destroy') {
+            resolve();
+          }
+          // Don't attempt to implement backpressure, since we're buffering the entire response anyway.
+          if (methodName !== 'write') {
+            return returnValue;
+          }
+        };
+      });
+      this.subject(req, res, function(err, _req, _res, _next) {
+        nextCalls.push(arguments);
+        if (nextCalls.length > 1) {
+          if (done) {
+            if (err) {
+              throw err;
+            } else {
+              throw new Error('next called more than once');
+            }
+          } else {
+            // Will be reported as a failure later
+            return;
+          }
+        }
+
+        // handle calling next() with a status code
+        if (typeof err === 'number') {
+          const statusCode = err;
+          err = new Error(`${statusCode}`);
+          err.statusCode = statusCode;
+          errorPassedToNext = err;
+        } else {
+          errorPassedToNext = err;
+        }
+
+        resolve();
+      });
+    })
+      .then(() => {
+        if (res.connection._writableState.corked > 0) {
+          // Wait for the connection to become uncorked before proceeding
+          const originalUncork = res.connection.uncork;
+          return new Promise(resolve => {
+            res.connection.uncork = function() {
+              const returnValue = originalUncork.apply(this, arguments);
+              if (res.connection._writableState.corked === 0) {
+                resolve();
+              }
+              return returnValue;
+            };
+          });
+        }
+      })
+      .then(() => {
+        Object.assign(context, {
+          req,
+          res,
+          metadata: {
+            strictAsync: isAsync,
+            errorPassedToNext: false,
+            isDestroyed,
+            requestDestroyed,
+            nextCalled: nextCalls.length > 0,
+            locals: res.locals,
+            url: req.url
+          },
+          httpRequest
+        });
+
+        let hasWrittenErrorStatusCode = false;
+        if (errorPassedToNext && !res.headersSent) {
```

Comment:
This is another chunk that is worth some special attention. Previously, setting the status code in the case of error was split between two blocks of code. Instead, we now make sure an error with a status code exists by the time we reach here (see https://github.com/unexpectedjs/unexpected-express/pull/88/files#diff-d913412612f980ff02cb5eb8fcc15f31R248) and then check error objects for an appropriate status code to use for the response.